pkg/cmd/pulumi/new.go

```diff
@@ -390,6 +451,205 @@ func useSpecifiedDir(dir string) (string, error) {
+	return conversationURL, err
+}
+
+func sendPromptToPulumiAI(
+	promptMessage string,
+	conversationID string,
+	connectionID string,
+	userName string,
+	language string,
+) (string, string, string, error) {
+	pulumiAIURL := env.AIServiceEndpoint.Value()
+	if pulumiAIURL == "" {
+		pulumiAIURL = "https://www.pulumi.com/ai"
+	}
+	parsedURL, err := url.Parse(pulumiAIURL)
+	if err != nil {
+		return "", "", "", err
+	}
+	requestPath := parsedURL.JoinPath("api", "chat")
+	requestBody := AIPromptRequestBody{
+		Language:       language,
+		Instructions:   promptMessage,
+		Model:          "gpt-4-turbo",
+		ResponseMode:   "code",
+		ConversationID: conversationID,
+		ConnectionID:   connectionID,
+	}
+	marshalledBody, err := json.Marshal(requestBody)
+	if err != nil {
+		return "", "", "", err
+	}
+	userDataCookie := http.Cookie{Name: "pulumi_command_line_user_name", Value: userName}
+	request, err := http.NewRequest("POST", requestPath.String(), bytes.NewReader(marshalledBody))
+	if err != nil {
+		return "", "", "", err
+	}
+	request.AddCookie(&userDataCookie)
+	fmt.Println("Sending prompt to Pulumi AI...")
+	res, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return "", "", "", err
+	}
+	defer res.Body.Close()
+	reader := bufio.NewReader(res.Body)
+	fmt.Println("Pulumi AI response:")
+	for {
+		chunk, err := reader.ReadByte()
+		if err != nil && err.Error() != "EOF" {
+			return "", "", "", err
+		}
+		fmt.Print(string(chunk))
+		if err != nil && err.Error() == "EOF" {
+			// Add a newline to ensure we don't overwrite the last line of the Pulumi AI response
+			fmt.Println()
+			break
+		}
+	}
+	conversationID = res.Header.Get("x-conversation-id")
+	connectionID = res.Header.Get("x-connection-id")
+	conversationURL := parsedURL.JoinPath("api", "project", url.PathEscape(fmt.Sprintf("%s.zip", conversationID))).String()
+	return conversationURL, connectionID, conversationID, nil
+}
+
+func runAINewPromptStep(
+	opts display.Options,
+	language string,
+	prompt string,
+	yes bool,
+	userName string,
+	currentContinueSelection string,
+	connectionID string,
+	conversationURL string,
+	conversationID string,
+) (continueSelection string,
+	conversationURLReturn string,
+	conversationIDReturn string,
+	connectionIDReturn string,
+	err error,
+) {
+	var promptMessage string
+	if prompt == "" || currentContinueSelection != "" {
+		if err := survey.AskOne(&survey.Input{
+			Message: "Please input your prompt here:\n",
+		}, &promptMessage, surveyIcons(opts.Color)); err != nil {
+			return "", "", "", "", err
+		}
+	} else {
+		promptMessage = prompt
+	}
+	conversationURLReturn, connectionIDReturn, conversationIDReturn, err = sendPromptToPulumiAI(
+		promptMessage,
+		conversationID,
+		connectionID,
+		userName,
+		language,
+	)
+	if err != nil {
+		return "", "", "", "", err
+	}
+	if connectionID != "" && connectionID != connectionIDReturn {
+		connectionIDReturn = connectionID
+		fmt.Println("Connection ID changed, please restart the prompt")
```

Comment:
What does this mean - and is it intentional to print it here?

Reply:
I've yet to reliably recreate an issue I saw earlier on in development - it can be the case where a "new session" is started if a request is missing user information, or if user information changes from one prompt to the next. This may leak some of the backend implementation, but when that happens, we generate a new "connection" object, which is what is used to determine how many follow-up prompts a user has made in the same conversation. It's also the metric we use to determine whether to prompt a user to log in to p-cloud after a number of unauthenticated prompts.